generators/chipyard/src/main/scala/ChipTop.scala

```diff
@@ -49,71 +52,121 @@ case object GlobalResetAsynchronousFull extends GlobalResetScheme with HasAsyncI
-  val systemReset = Wire(Input(Reset()))
 
   // The system module specified by BuildSystem
   val lSystem = p(BuildSystem)(p).suggestName("system")
-  val system = withClockAndReset(systemClock, systemReset) { Module(lSystem.module) }
-
-  // Call all of the IOBinders and provide them with a default clock and reset
-  withClockAndReset(systemClock, systemReset) {
-    // Call each IOBinder on both the lazyModule instance and the module
-    // instance. Generally, an IOBinder PF should only be defined on one, so
-    // this should not lead to two invocations.
-    val (_ports, _iocells, _harnessFunctions) = p(IOBinders).values.flatMap(f => f(lSystem) ++ f(system)).unzip3
-    // We ignore _ports for now...
-    iocells ++= _iocells.flatten
-    harnessFunctions ++= _harnessFunctions.flatten
-  }
 
+  // The provider of all clock groups. Subclasses define this
+  def provideDiplomaticClocks(): Unit
+
+  // This function generates RTL that may be necessary to drive the diplomatic clocks
+  // Subclasses define this
+  def driveClockGroups(systemClock: Clock, systemReset: Reset): (Seq[IOCell], Seq[TestHarnessFunction])
+
+
+  // Provides the implicit clock for the lSystem
+  val systemClockSinkNode = ClockSinkNode(List(ClockSinkParameters()))
+  val clockGroupNode = ClockGroupIdentityNode()
+
+  // Connect the lazy system's diplomatic clocks to ours
+  lSystem match { case l: BaseSubsystem => l.asyncClockGroupsNode :*= clockGroupNode }
+  systemClockSinkNode := ClockGroup()(p, ValName("chipyard_system")) := clockGroupNode
+
+  // Now that all the clock sinks are connected to the clockGroupNode, provide the
+  // clock source with this configurably-defined function
+  provideDiplomaticClocks()
+
+  lazy val module: LazyModuleImpLike = new LazyRawModuleImp(this) {
+    // The system clock
+    // These are given so that IOCell can use DataMirror and generate ports with
+    // The right flow (Input/Output). These do not generate IOs
+    val systemClock = Wire(Input(Clock()))
+    val systemReset = Wire(Input(Reset()))
+
+    // Call all of the IOBinders and provide them with a default clock and reset
+    withClockAndReset(systemClock, systemReset) {
+      // Call each IOBinder on both the lazyModule instance and the module
+      // instance. Generally, an IOBinder PF should only be defined on one, so
+      // this should not lead to two invocations.
+      val (_ports, _iocells, _harnessFunctions) = p(IOBinders).values.flatMap(f => f(lSystem) ++ f(lSystem.module)).unzip3
+      // We ignore _ports for now...
+      iocells ++= _iocells.flatten
+      harnessFunctions ++= _harnessFunctions.flatten
+    }
+
+    // Set the System clock/reset to the diplomatic clock bundle
+    systemClock := systemClockSinkNode.in.head._1.clock
+    systemReset := systemClockSinkNode.in.head._1.reset
+
+    // The ModuleImp might be a LazyRawModule, in which case there is no clock
+    lSystem.module match { case l: LazyModuleImp => {
```

Comment:
Is the idea that they would have a custom ChipTop or use binders to drive the clock of the lSystem themselves?

Reply:
I guess there's three ways to specify clock input.

1. As implicit Chisel clocks. The current setup supports this case.
2. As a `ClockNode` within the system. The current code matches `BaseSubsystem`, but I think it should more accurately match `HasConfigurablePRCILocations` to match any `System` which might provide diplomatic clocks.
3. As some other named IO. I hope that we can define some generic "ChipTopClockingKey" to provide enough flexibility to support these cases.